### PR TITLE
fix(ui): allow XML tag for markdown rich input

### LINKED ISSUE #6047: [3.1.1] HTML/XML tags are getting stripped from prompt
### Describe the bug

When saving a prompt that contains structured tags (e.g. `<instructions>`, `<context>`), the editor removes these tags, resulting in a flattened, unstructured prompt. This happens either in Rich-Text editor mode and Source editor mode. This is probably a side effect of the new editor implemented at https://github.com/FlowiseAI/Flowise/pull/5988, reverting back to 3.1.0 fixed the issue for me.

This is problematic because major LLM providers explicitly recommend structured prompting with XML-style tags. Removing them breaks the intended structure, reduces model reliability, and undermines best practices for clear instruction separation and predictable outputs.

Guidelines:
* https://platform.claude.com/docs/en/build-with-claude/prompt-engineering/claude-prompting-best-practices#structure-prompts-with-xml-tags
* https://developers.openai.com/api/docs/guides/prompt-engineering?prompt-example=prompt#message-formatting-with-markdown-and-xml

### To Reproduce

1. Open Agentflow V2
2. Create a new LLM/Agent node
3. Open any prompt input
4. Enter the following text:
```
<question>
{{question}}
</question>
<current_date_time>
{{current_date_time}}
</current_date_time>
```
5. Save and reopen

### Expected behavior

The tags should be retained in the prompt output and not being removed.

### Screenshots

Before saving:
Image

After saving:
Image

### Flow

_No response_

### Use Method

Docker

### Flowise Version

3.1.1

### Operating System

Linux

### Browser

Chrome

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jocelynlin-wd**
- Introduced utilities for escaping and unescaping non-standard XML/HTML tags to preserve them during markdown processing.
- Updated RichInput and ExpandRichInputDialog components to utilize the new XML tag utilities for handling custom tags.
- Added Jest configuration for the UI package and created tests for the XML tag utilities to ensure functionality.
- Updated package.json to include Jest as a dependency and added a test script.

---
flow data:

Before fix:
```
        "inputs": {
          "agentModel": "chatGoogleGenerativeAI",
          "agentMessages": [
            {
              "role": "assistant",
              "content": "You are a code review assistant. Analyze the pull request and provide structured feedback.\n\nPR URL: {{question}}\n\n## Summary Provide a 2-3 sentence overview.\n\n## Verdict\n\nAPPROVE, REQUEST_CHANGES, or COMMENT"
            }
          ],
          /// omit ...
        },
```

After fix:
```
        "inputs": {
          "agentModel": "chatGoogleGenerativeAI",
          "agentMessages": [
            {
              "role": "assistant",
              "content": "<instructions>\n\nYou are a code review assistant. Analyze the pull request and provide structured feedback.\n\n</instructions>\n\n<context>\n\nRepository: [https://github.com/FlowiseAI/Flowise](https://github.com/FlowiseAI/Flowise)\n\nPR: {{question}}\n\n</context>\n\n<output_format>\n\n## Summary\n\n- **PR**: [https://github.com/FlowiseAI/Flowise/pull/](https://github.com/FlowiseAI/Flowise/pull/{{question){{question}}[[](https://github.com/FlowiseAI/Flowise/pull/{{question)\n- **Overview**: Provide a 2-3 sentence overview of the changes.\n\n## Review\n\n- **Quality**: Rate code quality 1-5\n- **Risk**: Highlight any risky changes\n- **Suggestions**: List actionable improvements\n\n## Verdict\n\nAPPROVE, REQUEST_CHANGES, or COMMENT\n\n</output_format>"
            }
          ],
          ... omit ...
        },
```

mix xml with html
```
          "agentMessages": [
            {
              "role": "assistant",
              "content": "<instructions>\n\nWhen generating HTML, use <div> for containers and <span> for inline elements.\n\nFor example:\n\n```\n\n<body>\n\n    <div>Container</div>\n\n    <span>Inline Element</span>\n\n</body>\n\n```\n\nAlways wrap content in semantic tags.\n\n</instructions>\n\n<context>\n\nFramework: React\n\nPR URL: {{question}}\n\n</context>\n\n<output_format>\n\n## Review\n\nList any missing semantic HTML tags.\n\n</output_format>"
            }
          ]
```

---
Recordings

Before:

https://github.com/user-attachments/assets/18497911-04a7-4b97-aa19-31fd49c329a0

After:

https://github.com/user-attachments/assets/c245aa01-abe5-48b1-a282-51cab2357695

https://github.com/user-attachments/assets/738d9033-e2ae-412f-88ed-8361bcad6e8b

https://github.com/user-attachments/assets/f443bb5d-7782-4a06-be4a-86e07a31bed1

### CODE REVIEW

File: packages/ui/src/utils/xmlTagUtils.js [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ * ]}
+ * unescapeXmlEntities(json)
+ * // json.content[0].content[0].text → '<question>What?</question>'
+ *
+ * @param {object} json - ProseMirror document JSON from editor.getJSON()
+ * @returns {object} The same JSON with decoded entities in text nodes
+ */
+export function unescapeXmlEntities(json) {
+    if (json.text) {
+        json.text = json.text.replace(/&lt;/g, '<').replace(/&gt;/g, '>')
```
[the remark this answers is not in the document]
Comment (jocelynlin-wd) [reply packages/ui/src/utils/xmlTagUtils.js]: updated

File: packages/ui/src/utils/xmlTagUtils.js [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * unescapeCustomXmlTags('<question>text</question>')
+ * // → '<question>text</question>'  (raw tags pass through unchanged)
+ *
+ * @param {string} text - Markdown output from TipTap
+ * @returns {string} Content with non-standard tags restored to angle brackets
+ */
+export function unescapeCustomXmlTags(text) {
+    if (!text || typeof text !== 'string') return text
+    return text.replace(/&lt;(\/?)([a-zA-Z][a-zA-Z0-9_.-]*)(\s[^&]*)?(\/?)&gt;/g, (match, slash, tagName, attrs, selfClose) => {
```
[the remark this answers is not in the document]
Comment (jocelynlin-wd) [reply packages/ui/src/utils/xmlTagUtils.js]: updated

File: packages/ui/src/utils/xmlTagUtils.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    'font',
+    'big',
+    'tt'
+])
+
+/**
+ * Regex matching opening, closing, and self-closing XML/HTML tags.
+ * Captures: (1) optional slash, (2) tag name, (3) optional attributes, (4) optional self-close slash
+ */
+const XML_TAG_REGEX = /<(\/?)([a-zA-Z][a-zA-Z0-9_.-]*)(\s[^>]*)?(\/?)>/g
```
[the remark this answers is not in the document]
Comment (jocelynlin-wd) [reply packages/ui/src/utils/xmlTagUtils.js:38]: `.*? `doesn't help here either: both patterns stop at the first `>`. The only fix for `>` inside quoted attribute values is a much more complex regex. This edge case doesn't apply to prompt XML tags (e.g., <instructions>, <context>).

**Comment by HenryHengZJ** [comment]
since we are adding test, can we also add test for the markdown link? https://github.com/FlowiseAI/Flowise/pull/6061

and also the original issue: https://github.com/FlowiseAI/Flowise/issues/5527

i should've added this test when working on the PR, sorry!

File: packages/server/src/utils/buildAgentflow.ts [hunk 4]
```diff
@@ -247,8 +247,9 @@ export const resolveVariables = async (
         // If value is not a string, return as is
         if (typeof value !== 'string') return value
 
-        // Convert legacy HTML content to markdown, preserving any markdown syntax within
-        if (/<[a-z][a-z0-9]*[^>]*>/i.test(value)) {
+        // Convert legacy HTML content to markdown, preserving any markdown syntax within.
+        // Only match standard HTML tags to avoid stripping intentional XML tags like <question>, <context> etc.
+        if (/<(?:p|div|span|h[1-6]|ul|ol|li|br|code|pre|blockquote|table|strong|em)\b[^>]*>/i.test(value)) {
```
Comment (HenryHengZJ) [inline packages/server/src/utils/buildAgentflow.ts]: how do determine if the user is actually having the html code inside the prompt?

for example, user might have something like:

```
<instruction>
Help me debug this code:
<body>
   <div>Test</div>
</body>
</instruction>
```

Comment (jocelynlin-wd) [reply packages/server/src/utils/buildAgentflow.ts]: do we want to support html tags as well? let me take a look

Comment (HenryHengZJ) [reply packages/server/src/utils/buildAgentflow.ts]: it might be rare but technically user can actually have any kind of prompt right?

Comment (jocelynlin-wd) [reply packages/server/src/utils/buildAgentflow.ts]: yes makes sense, updated to support html. Here is the recording:

https://github.com/user-attachments/assets/f443bb5d-7782-4a06-be4a-86e07a31bed1

prompt:
````
<instructions>
When generating HTML, use <div> for containers and <span> for inline elements.

For example:
```
<body>
    <div>Container</div>
    <span>Inline Element</span>
</body>
```
Always wrap content in semantic tags.
</instructions>

<context>
Framework: React
PR URL: {{question}}
</context>

<output_format>
## Review
List any missing semantic HTML tags.
</output_format>
````
and flow data:
```
          "agentMessages": [
            {
              "role": "assistant",
              "content": "<instructions>\n\nWhen generating HTML, use <div> for containers and <span> for inline elements.\n\nFor example:\n\n```\n\n<body>\n\n    <div>Container</div>\n\n    <span>Inline Element</span>\n\n</body>\n\n```\n\nAlways wrap content in semantic tags.\n\n</instructions>\n\n<context>\n\nFramework: React\n\nPR URL: {{question}}\n\n</context>\n\n<output_format>\n\n## Review\n\nList any missing semantic HTML tags.\n\n</output_format>"
            }
          ],
```

**Comment by jocelynlin-wd** [comment]
> since we are adding test, can we also add test for the markdown link? #6061
> 
> and also the original issue: #5527
> 
> i should've added this test when working on the PR, sorry!

@HenryHengZJ  I added test case to cover markdown format (including links), but for the actual changes in #6061 we will need to do react component test which require more setups etc. I will add test in separate PR instead. (same for the original issue in #5527 )

**Comment by HenryHengZJ** [comment]
when I run `pnpm build` I get this error:

```
flowise-ui:build: Error: [vite]: Rollup failed to resolve import "@tiptap/core" from "C:/Users/Henry/Codes/Flowise/packages/ui/src/ui-component/input/RichInput.jsx".
flowise-ui:build: This is most likely unintended because it can break your application at runtime....
```

Looks like it fails because Rollup cannot resolve the `@tiptap/core` import in `RichInput.jsx`. The module is used (line 6: import { mergeAttributes } from '@tiptap/core') but not declared as a direct dependency — it only exists as a transitive dep of other @tiptap/* packages.

We might need to add `@tiptap/core` to packages/ui/package.json dependencies at version 3.20.4 (matching the other tiptap packages)

File: packages/server/src/utils/buildAgentflow.ts (hunk 4 above)
Comment (HenryHengZJ) [reply packages/server/src/utils/buildAgentflow.ts]: the line break seems to be off:
image

Comment (HenryHengZJ) [reply packages/server/src/utils/buildAgentflow.ts]: I think it might be just a rendering issue on the Process Flow side. I used langsmith and console to check it out seems to be sending correctly:

image


image

image

**Review by HenryHengZJ — APPROVED** [review]
